### ISSUE TITLE
Different results on same test dataset 

### ISSUE THREAD

**Issue by mansi-aggarwal-2504** [ultralytics/yolov5#3516]
I trained a model and was testing it on a bunch of images using:
`!python test.py --weights '/content/gdrive/MyDrive/YOLOV5_TRAINING/yolov5/runs/train/exp6/weights/best.pt' --data '/content/gdrive/MyDrive/YOLOV5_TRAINING/data.yaml' --task=test`
Then, the resulting confusion matrix was
![confusion_matrix](https://user-images.githubusercontent.com/60866597/121138879-c5b9de00-c855-11eb-84fa-def42d9a3238.png)

I tested the same dataset again without changing anything and then results dropped significantly suddenly:
![confusion_matrix-3](https://user-images.githubusercontent.com/60866597/121139087-0154a800-c856-11eb-8e9b-ba28538fc3bb.png)

I thought there might be some error on my end due to wrong path or something, so I trained the model again using the same dataset and I verified the runs/train/exp confusion matrix with my old train run and it was exactly the same. Now, I tested on the same bunch of images and it's giving me 76% TP again. 
I don't understand the reason behind this. A little variation would've been okay, but the training results are exactly the same. So it should've been the same on the test set also.

**Comment by glenn-jocher** [comment]
@mansi-aggarwal-2504 testing is a deterministic operations, so if your results changed this would be directly correlated to a change in your data, code, or model.

**Comment by mansi-aggarwal-2504** [comment]
I am positive I didn't make any changes to the data or the code. I am not customising any parameters while running `test.py` either. Just providing the path to my weights i.e. best.pt and test dataset.
Results on training set are exactly the same. That's why I don't understand such a huge difference on test dataset.
I also ran `detect.py` again on the test set. The number of objects detected in each image are also slightly more after the second run which is a bit odd because results are changing on the same dataset.

**Comment by glenn-jocher** [comment]
@mansi-aggarwal-2504 detect.py and test.py are deterministic operations, so if your results changed this would be directly correlated to a change in your data, code, or model.

**Comment by mansi-aggarwal-2504** [comment]
Hi @glenn-jocher, I kept all parameters same in `test` and `detect`.
test.py: `Namespace(augment=True, batch_size=1, conf_thres=0.35, data='data.yaml', device='', exist_ok=False, img_size=1280, iou_thres=0.45, name='exp', project='runs/test', save_conf=True, save_hybrid=True, save_json=False, save_txt=True, single_cls=False, task='test', verbose=False, weights=['<xyz>/runs/train/exp7/weights/best.pt'])
`

(path of data accessed by test.py using data.yaml: `colour_test/images/test`)

detect.py: `Namespace(agnostic_nms=False, augment=True, classes=None, conf_thres=0.35, device='', exist_ok=False, hide_conf=False, hide_labels=False, img_size=1280, iou_thres=0.45, line_thickness=3, name='exp', nosave=False, project='runs/detect', save_conf=True, save_crop=False, save_txt=True, source='colour_test/images/test', update=False, view_img=False, weights=['<xyz>/runs/train/exp7/weights/best.pt'])`


The main reason behind this activity is to get the confusion matrix for my predictions. I first get my predictions using `detect`. Now I want to check how good the predictions are. So I supply the ground truth and run `test.py` using the same parameters but somehow, the results are different. So I am not finding a way to check my results. Any
way, I could reproduce the same results as `detect.py` and obtain the confusion matrix as well?

**Comment by glenn-jocher** [comment]
@mansi-aggarwal-2504 👋 Hello, thank you for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [test.py](https://github.com/ultralytics/yolov5/blob/master/test.py) in YOLOv5.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, test.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:

## `train.py`

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L188-L192
- **testloader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L199-L202
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## `test.py`

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/test.py#L89-L90
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`

## `detect.py`

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/detect.py#L46-L53
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`

## YOLOv5 PyTorch Hub

`models.autoShape()` class used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/models/common.py#L225-L250

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

**Comment by mansi-aggarwal-2504** [comment]
Hi @glenn-jocher, thank you for explaining the differences between the three modules. 
So, this is what I am trying instead: since I want an evaluation of my results and confusion matrix can be a good starting point, I tried to integrate that module in `detect.py` 

- I also read images and labels here using `create_dataloader`
- I have added another parameter to the parser so that the code knows if ground truth labels are available or not:
`parser.add_argument('--groundtruth', default=False, action='store_true', help='get confusion matrix using ground truth labels')`
- Calculating the confusion matrix:
 ```
      if opt.groundtruth:
                labels = targets[targets[:, 0] == i, 1:]
                tbox = xywh2xyxy(labels[:, 1:5])
                confusion_matrix.process_batch(predn, torch.cat((labels[:, 0:1], tbox), 1))
```
The variables defined in the above part have been defined in the same way as they are in `test.py`

The problem is my confusion matrix is coming out to be wrong:
![confusion_matrix](https://user-images.githubusercontent.com/60866597/123193168-e7f06480-d4c1-11eb-8600-63c23e9c95bf.png)

Even if somehow none of my predictions matched as True Positives, I have zero False Positives which is suspicious. 
https://github.com/ultralytics/yolov5/blob/ae4261c7749ff644f45c66b79ecb1fff06437052/utils/metrics.py#L128-L131

In line 131 here, the `iou` is coming out to be zeros tensor. In `metric.py`, the `inter` becomes zero when `clamp(0)` is applied.

https://github.com/ultralytics/yolov5/blob/ae4261c7749ff644f45c66b79ecb1fff06437052/utils/general.py#L513

What is [clamp(0)](https://www.geeksforgeeks.org/python-pytorch-clamp-method/) doing here and why has the parameter set to zero? 
I don't understand what I am doing wrong. I would be happy to share my modified `detect.py` if that helps.
